### PR TITLE
add more lighttpd code to static analysis

### PR DESCRIPTION
**Pull request by aneksteind**

### CODE REVIEW

File: analysis/tests/lighttpd-minimal/src/main.rs [hunk 1]
```diff
@@ -4,88 +4,1149 @@
...
 pub struct server {
+    pub plugin_slots: *mut libc::c_void,
+    pub config_context: *mut array,
+    pub config_captures: libc::c_int,
     pub ev: *mut fdevents,
-    pub cur_fds: libc::c_int,
+    pub network_backend_write: Option<
+        unsafe extern "C" fn(libc::c_int, *mut chunkqueue, off_t, *mut log_error_st) -> libc::c_int,
+    >,
+    pub request_env: Option<unsafe extern "C" fn(*mut request_st) -> handler_t>,
```
Comment (kkysen) [inline analysis/tests/lighttpd-minimal/src/main.rs:231]: Is leaving these in types fine?

File: c2rust-analyze/src/main.rs [hunk 2]
```diff
@@ -742,7 +742,7 @@ fn fn_body_owners_postorder(tcx: TyCtxt) -> Vec<LocalDefId> {
 
         match tcx.def_kind(root_ldid) {
             DefKind::Fn | DefKind::AssocFn => {}
-            DefKind::AnonConst => continue,
+            DefKind::AnonConst | DefKind::Const => continue,
```
Comment (kkysen) [inline c2rust-analyze/src/main.rs:745]: What does this fix?

File: analysis/tests/lighttpd-minimal/src/main.rs [hunk 3]
```diff
@@ -183,8 +1292,14 @@ unsafe extern "C" fn fdnode_free(fdn: *mut libc::c_void) {
 }
 
 pub unsafe extern "C" fn lighttpd_test() {
-    let fdarr = malloc(::std::mem::size_of::<*mut fdnode>() as libc::c_ulong); // TODO: handle cast as *mut *mut fdnode;
-    let fdes = malloc(::std::mem::size_of::<fdevents>() as libc::c_ulong); // TODO: handle cast as *mut fdevents;
+    let fdarr = calloc(
+        0 as libc::c_int as libc::c_ulong,
```
Comment (kkysen) [inline analysis/tests/lighttpd-minimal/src/main.rs:1296]: Can you just confirm these are supposed to be 0?  I remember we had a nullptr deref from something like this before.

**Review by kkysen — COMMENTED** [review]
I think this should all be fine as long as it passes the tests.  That said, it's a bit hard to see what exactly changed here since the automatic diff isn't very clean, so it'd be helpful to separate this into a few commits (e.x. all the new function declarations and types, and then the function body changes) (I can take a look at this).

File: analysis/tests/lighttpd-minimal/src/main.rs [hunk 4]
```diff
@@ -4,88 +4,1149 @@
...
         (*srv).ev,
         (*con).fd,
         // Some(
         //     connection_handle_fdevent
         //         as unsafe extern "C" fn(*mut libc::c_void, libc::c_int) -> handler_t,
         // ),
         // con as *mut libc::c_void,
-        fdn
+        // con as *mut libc::c_void,
+        fdn,
```
Comment (kkysen) [inline analysis/tests/lighttpd-minimal/src/main.rs:1062]: I think this duplicates the `con` arg.

**Comment by kkysen** [comment]
> I think this should all be fine as long as it passes the tests. That said, it's a bit hard to see what exactly changed here since the automatic diff isn't very clean, so it'd be helpful to separate this into a few commits (e.x. all the new function declarations and types, and then the function body changes) (I can take a look at this).

I opened #792 for this for how I think it'd be better to split up the commits.

File: analysis/tests/lighttpd-minimal/src/main.rs (hunk 1 above)
Comment (aneksteind) [reply analysis/tests/lighttpd-minimal/src/main.rs:231]: yes, although I'm not sure what you're concerned about

File: c2rust-analyze/src/main.rs (hunk 2 above)
Comment (aneksteind) [reply c2rust-analyze/src/main.rs:745]: being able to handle things like `pub const HANDLER_GO_ON: handler_t = 0;` without associating a body with them

**Comment by aneksteind** [comment]
closing in favor of #792

File: analysis/tests/lighttpd-minimal/src/main.rs (hunk 3 above)
Comment (aneksteind) [reply analysis/tests/lighttpd-minimal/src/main.rs:1296]: these are contrived for the time being, the actual code is more involved

Comment (kkysen) [reply analysis/tests/lighttpd-minimal/src/main.rs:1296]: That's fine as long as we make sure to update them before they are used and not just immediately freed.

File: analysis/tests/lighttpd-minimal/src/main.rs (hunk 4 above)
Comment (kkysen) [reply analysis/tests/lighttpd-minimal/src/main.rs:1062]: I fixed this in #792.

File: analysis/tests/lighttpd-minimal/src/main.rs (hunk 1 above)
Comment (kkysen) [reply analysis/tests/lighttpd-minimal/src/main.rs:231]: I just know we can't fully handle function pointers, so I wasn't sure if it's okay to have them in types and just not use them.

Comment (aneksteind) [reply analysis/tests/lighttpd-minimal/src/main.rs:231]: that's correct, only a subset of types are iterated over (ADTs) and it's their use in MIR bodies that would pose an issue

**Pull request closed without merging**